proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+
+The VC manager should initiate the presentation (send the RPC request) and store the VC pair in a private list. It should be possible to associate RPC responses to the VC pair in the list. If a response is received and can be associated to a VC pair then the VC pair should be removed from the list and the completion handler should be called.
+
+The oldest VC pair of this list should be the currently presenting and presented VC. The currently presenting VC will be accessible through the `.presentingModalViewController` property and the presented VC through the `.presentedModalViewController`. If the private list is empty the two properties should return nil.
+
+By the nature of the HMI overlays can be aborted by another overlay or get rejected because of a currently presented overlay (or something of a higher priority e.g. VR session). The private list offers flexibility in case the app calls the method multiple times and can deal with rejected overlays that were never visible and aborted overlays if another one is requested.
+
+#### SDLModalViewController
+
+This class should be used as the base class for any kind of overlay. As it inherits `SDLViewController` it is possible to manipulate the head unit during a modal presentation.
+
+```objc
+@interface SDLModalViewController : SDLViewController
+
+@property (nonatomic, assign) NSTimeInterval duration; // used throughout all subclasses
+@property (nonatomic, copy) NSString *message; // used throughout all subclasses
+
+- (void)present:(SDLApplication *)application; // private method called by the VC manager
+
+@end
+```
+
+#### SDLAlertController
+
+The alert controller is used to abstract the RPC `Alert`. The properties match the parameters of the RPC.
+
+```objc
+@interface SDLAlertController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@property (nonatomic) BOOL shouldPlayTone;
+@property (nonatomic) BOOL progressIndicator;
+
+@end
+``` 
+
+#### SDLAudioInputController
+
+The audio input controller is used to abstract the RPC `PerformAudioPassThru`, `OnAudioPassThru` and `EndAudioPassThru`. The app developer should be able to provide a block which is called for audio data. Furthermore a method called `dismiss` should help to end the audio pass thru.
+
+```objc
+@interface SDLAudioInputController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) SDLSamplingRate samplingRate;
+@property (nonatomic, copy) SDLBitsPerSample bitsPerSample;
+@property (nonatomic, copy) SDLAudioType audioType;
+@property (nonatomic) BOOL muteMediaSource;
+
+@property (nonatomic) (^onAudioData)(NSData *audioData);
```

Comment:
`onAudioData` seems perhaps a bit out of place compared to the rest of this API and iOS' naming.